misc/python/materialize/parallel_workload/parallel_workload.py

```diff
@@ -0,0 +1,302 @@
+        cur.execute("ALTER SYSTEM SET max_roles = 105")
+        cur.execute("ALTER SYSTEM SET max_clusters = 105")
+        cur.execute("ALTER SYSTEM SET max_replicas_per_cluster = 105")
+
+    end_time = (
+        datetime.datetime.now() + datetime.timedelta(seconds=runtime)
+    ).timestamp()
+
+    rng = random.Random(random.randrange(SEED_RANGE))
+    database = Database(rng, seed, host, port, system_port, complexity, scenario)
+    conn = pg8000.connect(host=host, port=port, user="materialize")
+    conn.autocommit = True
+    with conn.cursor() as cur:
+        database.create(Executor(rng, cur))
+    conn.close()
+
+    conn = pg8000.connect(
+        host=host, port=port, user="materialize", database=str(database)
+    )
+    conn.autocommit = True
+    with conn.cursor() as cur:
+        database.create_relations(Executor(rng, cur))
+    conn.close()
+
+    workers = []
+    threads = []
+    for i in range(num_threads):
+        worker_rng = random.Random(rng.randrange(SEED_RANGE))
+        weights: List[float]
+        if complexity == "ddl":
+            weights = [60, 30, 30, 30, 10]
+        elif complexity == "dml":
+            weights = [60, 30, 30, 30, 0]
+        elif complexity == "read":
+            weights = [60, 30, 0, 0, 0]
+        else:
+            raise ValueError(f"Unknown complexity {complexity}")
+        action_list = worker_rng.choices(
+            [
+                read_action_list,
+                fetch_action_list,
+                write_action_list,
+                dml_nontrans_action_list,
+                ddl_action_list,
+            ],
+            weights,
+        )[0]
+        actions = [
+            action_class(worker_rng, database)
+            for action_class in action_list.action_classes
+        ]
+        worker = Worker(
+            worker_rng,
+            actions,
+            action_list.weights,
+            end_time,
+            action_list.autocommit,
+            system=False,
+        )
+        thread_name = f"worker_{i}"
+        print(
+            f"{thread_name}: {', '.join(action_class.__name__ for action_class in action_list.action_classes)}"
+        )
+        workers.append(worker)
+
+        thread = threading.Thread(
+            name=thread_name,
+            target=worker.run,
+            args=(host, port, "materialize", str(database)),
+        )
+        thread.start()
+        threads.append(thread)
+
+    if scenario == "cancel":
+        worker = Worker(
+            worker_rng,
+            [CancelAction(worker_rng, database, workers)],
+            [1],
+            end_time,
+            autocommit=False,
+            system=True,
+        )
+        workers.append(worker)
+        thread = threading.Thread(
+            name="cancel",
+            target=worker.run,
+            args=(host, system_port, "mz_system", str(database)),
+        )
+        thread.start()
+        threads.append(thread)
+    elif scenario == "kill":
```

Comment:
Same as for the complexities, scenarios should also be an ENUM or a class hierarchy.